src/cov/cov-exclude.c

```diff
@@ -30,101 +30,323 @@ struct _cover_exclude_ctx {
+                                               sizeof(cover_fold_cmd_t));
+            }
+            data->ef->fold[n].loc = loc;
+            data->ef->fold[n].found_target = false;
+            data->ef->fold[n].found_source = false;
+            data->ef->fold[n].target = ident_new(target);
+            data->ef->fold[n].source = ident_new(source);
+            data->ef->n_fold_cmds++;
          }
          else
-            fatal_at(&ctx.loc, "invalid command: $bold$%s$$", tok);
+            error_at(&loc, "invalid command: $bold$%s$$", tok);
 
          tok = strtok(NULL, delim);
       }
       line_num++;
    }
 
-   fclose(ctx.ef);
+   fclose(ef);
+}
+
+static void cover_apply_exclude_cmds(cover_data_t *data)
+{
+   assert (data->ef != NULL);
+
+   for (int i = 0; i < data->ef->n_excl_cmds; i++)
+      data->ef->excl[i].found = false;
+
+   cover_exclude_scope(data, data->root_scope);
+
+   for (int i = 0; i < data->ef->n_excl_cmds; i++)
+      if (!data->ef->excl[i].found)
+         warn_at(&data->ef->excl[i].loc, "excluded hierarchy does not match any "
+                 "coverage item: '%s'", istr(data->ef->excl[i].hier));
+}
+
+static void cover_fold_scopes(cover_scope_t *tgt_scope, cover_scope_t *src_scope)
+{
+   int tgt_prefix_len = ident_len(tgt_scope->hier);
+   int src_prefix_len = ident_len(src_scope->hier);
+
+   // Process items
+   for (int i = 0; i < src_scope->items.count; i++) {
+      cover_item_t *src = AREF(src_scope->items, i);
+
+      for (int j = 0; j < tgt_scope->items.count; j++) {
+         cover_item_t *tgt = AREF(tgt_scope->items, j);
+         ident_t tgt_suffix_hier = NULL;
+         ident_t src_suffix_hier = NULL;
+
+         // Compare hierarchical paths, but strip "tgt_scope" prefix from "tgt",
+         // and "src_scope" from "src". Only hierarchy suffix needs to be the same.
+         if (ident_len(tgt->hier) > tgt_prefix_len)
+            tgt_suffix_hier = ident_new(istr(tgt->hier) + tgt_prefix_len);
+         if (ident_len(src->hier) > src_prefix_len)
+            src_suffix_hier = ident_new(istr(src->hier) + src_prefix_len);
+
+         if ((tgt_suffix_hier == src_suffix_hier) && (tgt->flags == src->flags)) {
+
+#ifdef COVER_DEBUG_FOLD
+            printf("Folding coverage item:\n"
+                   "    tgt_prefix_len:      %d\n"
+                   "    src_prefix_len:      %d\n"
+                   "    tgt->hier:           %s\n"
+                   "    src->hier:           %s\n"
+                   "    tgt_suffix_hier:     %s\n"
+                   "    src_suffix_hier:     %s\n"
+                   "    src->data:           %d\n"
+                   "    tgt->data:           %d\n",
+                   tgt_prefix_len,
+                   src_prefix_len,
+                   istr(tgt->hier),
+                   istr(src->hier),
+                   istr(tgt_suffix_hier),
+                   istr(src_suffix_hier),
+                   src->data,
+                   tgt->data);
+#endif
```

Comment:
Can you get rid of these other COVER_DEBUG_FOLD blocks too please.

Reply:
I did. But I have the feeling I will need them in future. I can imagine the whole folding with the double
traversal of the scope hierarchy can have corner-cases that may have bugs...

Reply:
It's not much effort to add them back in if/when needed.